go/glome/glome.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package glome implements GLOME protocol.
+package glome
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"crypto/curve25519"
+	"crypto/hmac"
+
+	"golang.org/x/crypto/curve25519"
+)
+
+const (
+	// PrivateKeySize is the size of a PrivateKey in bytes.
+	PrivateKeySize = 32
+	// PublicKeySize is the size of a PublicKey in bytes.
+	PublicKeySize = 32
+	// SharedSecretSize is the size of a SharedSecret in bytes.
+	SharedSecretSize = 32
+	// Maximum size allow for a Tag
+	MaxTagSize = 32
+	// Minimum size allow for a Tag
+	MinTagSize = 1
+)
+
+var (
+	ErrInvalidPublicKey = errors.New(
+		fmt.Sprintf("glome: invalid public key - byte slice len is not %d", PublicKeySize))
+	ErrInvalidPrivateKey = errors.New(
+		fmt.Sprintf("glome: invalid private key - byte slice len is not %d", PrivateKeySize))
+	ErrInvalidTagSize = errors.New(
+		fmt.Sprintf("glome: invalid tag size - MinPeerTagSize must be in range %d-%d",
+			MinTagSize, MaxTagSize))
+)
+
+// SharedSecret is the type of GLOME (and x25519) shared secret.
+type SharedSecret [SharedSecretSize]byte
```

Comment:
I have not been a fan of fixed size byte slices and I have rejoiced when `curve25519.X25519` was added, which takes `[]byte` instead of `[32]byte`. The latter are just unwieldy to work with, as you need `[:]` everywhere and need to copy things.

The other question is if you want to make the type opaque. If so (and I'm not sure you want to), you'd do `type SharedSecret struct{ secret []byte }`, which makes the member inaccessible.

Reply:
Decided to delete this type.